Получение данных из форм

delannushka opened the pull request:
---
:mortar_board: [Получение данных из форм](https://up.htmlacademy.ru/php/14/user/2069609/tasks/13)

delannushka committed:
1. Если не вводить название задачи - валидация работает. Но добавить задачу не удается - ошибка

2. Валидация по дате работает наоборот, тк когда в функции меняю значения return - все ломается.
Комментарии дам отдельно.

delannushka committed:
1. Если не вводить название задачи - валидация работает. Но добавить задачу не удается - ошибка

2. Валидация по дате работает наоборот, тк когда в функции меняю значения return - все ломается.
Комментарии дам отдельно.

delannushka committed:
Не получается найти пуллреквест

1. Если не вводить название задачи - валидация работает. Но добавить задачу не удается - ошибка

2. Валидация по дате работает наоборот, тк когда в функции меняю значения return - все ломается.
Комментарии дам отдельно.

delannushka committed:
Исправила название ветки
1. Если не вводить название задачи - валидация работает. Но добавить задачу не удается - ошибка

2. Валидация по дате работает наоборот, тк когда в функции меняю значения return - все ломается.
Комментарии дам отдельно.

nkolosov commented on add.php:
```diff
@@ -0,0 +1,110 @@
+<?php
+
+function count_of_tasks($array_of_task, $id_of_category) {
```
Если одна и та же функция используется в нескольких сценариях, то нужно вынести функцию в отдельный файл и подключать через require/include

nkolosov commented on add.php:
```diff
@@ -0,0 +1,110 @@
+    $content = include_template('error.php', ['error' => $error]);
+}
+else {
+    $sql = 'SELECT * FROM projects WHERE user_id = ?';
```
Аналогично, если список проектов нужен на разных страницах, то получение списка проектов необходимо вынести в отдельный php-файл и подключать его где необходимо

nkolosov commented on add.php:
```diff
@@ -0,0 +1,110 @@
+        if (!empty($_FILES['file_txt']['name'])) {
+            $tmp_name = $_FILES['file_txt']['tmp_name'];
+            $link_to_file = $_FILES['file_txt']['name'];
+            $filename = uniqid() . ' .txt';
```
Мы загружаем различные типы файлов (например, фото), а тут всем добавляется расширение, как будто они текстовые файлы

nkolosov commented on add.php:
```diff
@@ -0,0 +1,110 @@
+        else {
+            // !!!!!ВОТ ЗДЕСЬ НЕ ПОНИМАЮ В ЧЕМ ОШИБКА: Data truncated for column 'project_id' at row 1
+            $sql = 'INSERT INTO tasks (date_of_create, name, link_to_file, due_date, user_id, project_id) VALUES (NOW(), ?, ?, ?, 3, ?)';
+            $stmt = db_get_prepare_stmt($link, $sql, [$task]);
```
Эта конструкция не будет работать, т.к `$task` не гарантирует необходимый порядок элементов. Вместо массива, необходимо передать отдельно все значения для placeholder'ов (?) в запросе

nkolosov commented on add.php:
```diff
@@ -0,0 +1,110 @@
+
+        $rules = [
+            'project_id' => function ($value) use ($projects_ids) {
+                return validateCategory($value, $projects_ids);
```
Функции стоит именовать по аналогии с другими функциями, в snake_case (а не camelCase)

nkolosov commented on templates/form_task.php:
```diff
@@ -0,0 +1,71 @@
+                        <select class="form__input form__input--select <?= $classname; ?>" name="project_id" id="project_id">
+                            <?php foreach ($projects as $project): ?>
+                                <option value="<?= $project['id'] ?>"
+                                    <?php if ($project['id'] == filter_input(INPUT_POST, 'project_id')): ?>selected<?php endif; ?>><?=$project['name'] ?></option>
```
Очень громоздкое сравнение получается, если бы введенный ранее проект просто был переменной, можно было бы сократить до `<?= $project['id'] === get_value('project_id') ? "selected" : "" ?>`

nkolosov commented on templates/form_task.php:
```diff
@@ -0,0 +1,71 @@
+                    <div class="form__row">
+                        <label class="form__label" for="date">Дата выполнения</label>
+                        <?php $classname = isset($errors['due_date']) ? "form__input--error" : ""; ?>
+                        <input class="form__input form__input--date <?= $classname; ?>" type="text" name="due_date" id="date" value="" placeholder="Введите дату в формате ГГГГ-ММ-ДД">
```
Здесь теряется ранее введенная дата

delannushka committed:
Внесла правки!

delannushka committed:
Внесла правки! и Добавила сортировку для вывода списка задач

delannushka committed:
Внесла правки! и Добавила сортировку для вывода списка задач

nkolosov commented on add.php:
```diff
@@ -0,0 +1,108 @@
+
+    $content = include_template('form_task.php', ['tasks' => $tasks, 'projects' => $projects]);
+
+    if ($_SERVER['REQUEST_METHOD'] == 'POST') {
```
Строгое сравнение всегда и везде (`===`)

nkolosov commented on add.php:
```diff
@@ -0,0 +1,108 @@
+        }
+        $errors = array_filter($errors);
+        if ($task['due_date']){
+            if (!is_date_valid($task['due_date'])){
```
Проверка должна быть в rules, `return condition1_valid() && condition2_valid()` условно

nkolosov commented on add.php:
```diff
@@ -0,0 +1,108 @@
+            $content = include_template('form_task.php', ['tasks' => $tasks, 'errors' => $errors, 'projects' => $projects]);
+        }
+        else {
+            $sql = 'INSERT INTO tasks (date_of_create, name, link_to_file, due_date, user_id, project_id) VALUES (NOW(), ?, ?, ?, 3, ?)';
```
3 тоже должног передаваться как параметр, current_user_id

nkolosov commented on helpers.php:
```diff
@@ -142,3 +142,35 @@ function include_template($name, array $data = []) {
+function is_date_correct($date) {
+    if ($date) {
+        $current_time = time();
+        define ('SECONDS_IN_DAY' , 86400);
```
Такая константа уже есть в index.php

nkolosov commented on templates/form_task.php:
```diff
@@ -0,0 +1,71 @@
+                    </ul>
+                </nav>
+
+                <a class="button button--transparent button--plus content__side-button" href="form-project.html">Добавить проект</a>
```
Ссылка на вёрстку ведет, а не на скрипт

nkolosov commented on templates/main.php:
```diff
@@ -77,15 +77,18 @@
             <td class="task__file">
-                <a class="download-link" href="#">Home.psd</a>
+                <?php if ($task['path']): ?>
+                    <a class="download-link" href="/uploads/<?=$task['path']; ?>">File.txt</a>
```
Тут надо отображать имя файла

nkolosov commented on add.php:
```diff
@@ -0,0 +1,108 @@
+            $link_to_file = $_FILES['file']['name'];
+            $filename = uniqid();
+            move_uploaded_file($tmp_name, 'uploads/' . $filename);
+            $task['link_to_file'] = $filename;
```
Теперь файлы получаются вообще без расширения.
А нужно сохранять исходное расширение файла

delannushka committed:
1. Когда нажимаю файл, на странице списка задач, открывается ссылка, но сам файл не загружается. И раньше файлы подгружались в папку uploads, а сейчас почему-то перестали. (из-за этого ссылки и перестали открываться)

2. добавила оба правила для 'due_date', но название ошибки перестало отображаться, при внесении некорреткной даты на странице Добавить Задачу.

nkolosov commented on helpers.php:
```diff
@@ -142,3 +122,45 @@ function include_template($name, array $data = []) {
+function is_date_correct($date) {
+    if ($date) {
+        $current_time = time();
+        define ('SECONDS_IN_ONE_DAY' , 86400);
```
Контанты определяются глобально, нельзя их определять внутри функции.

Лучше определение всех констант вынести в инициализацию проекта

nkolosov commented on list_of_projects.php:
```diff
@@ -0,0 +1,5 @@
+$sql = 'SELECT * FROM projects WHERE user_id = ?';
+$stmt = db_get_prepare_stmt($link, $sql, [$current_user_id]);
+mysqli_stmt_execute($stmt);
+$result = mysqli_stmt_get_result($stmt);
```
Вот тут стоит использовать какое-то другое имя для переменной, т.к. `result` встречается часто и при подключении разхных файлов одно значение может перетереть другое

nkolosov commented on add.php:
```diff
@@ -0,0 +1,104 @@
+            $content = include_template('form_task.php', ['tasks' => $tasks, 'errors' => $errors, 'projects' => $projects]);
+        }
+        else {
+            $sql = 'INSERT INTO tasks (date_of_create, name, link_to_file, due_date, user_id, project_id) VALUES (NOW(), ?, ?, ?, '. $current_user_id .', ?)';
```
current_user_id стоит тоже передавать через переменную prepared statement

nkolosov commented on add.php:
```diff
@@ -0,0 +1,104 @@
+
+        $rules = [
+            'project_id' => function ($value) use ($projects_ids) {
+                return validate_category($value, $projects_ids);
```
Вроде договорились, что это не категории, а проекты

nkolosov commented on add.php:
```diff
@@ -0,0 +1,104 @@
+                return validate_category($value, $projects_ids);
+            },
+            'due_date' => function($value) {
+                return is_date_correct($value) && is_date_valid($value);
```
Логичнее в функциях возвращать `true`/`false`, а тут, если ошибка, возвращать её текст.
Типа 
```
if (is_error1()) {
  return "error 1";
}

if (is_error2()) {
  return "error 2";
}

return null;
```

delannushka committed:
Внесла правки!

nkolosov commented on authorization.php:
```diff
@@ -0,0 +1,73 @@
+<?php
```
Это и другой задачи

nkolosov commented on helpers.php:
```diff
@@ -1,24 +1,4 @@
- *
- * @return bool true при совпадении с форматом 'ГГГГ-ММ-ДД', иначе false
- */
-function is_date_valid(string $date) : bool {
```
А зачем перенесла эту функцию отсюда? Не нужно уже готовые функции трогать и куда-то переносить

nkolosov commented on sign-up.php:
```diff
@@ -0,0 +1,56 @@
+<?php
```
Из другого задания

nkolosov commented on templates/form-authorization.php:
```diff
@@ -0,0 +1,46 @@
+<div class="content">
```
Из другого задания

nkolosov commented on templates/form-project.php:
```diff
@@ -0,0 +1,57 @@
+    <div class="content">
```
Из другого задания

nkolosov commented on templates/layout.php:
```diff
@@ -20,13 +20,13 @@
                             <p>Константин</p>
 
-                            <a href="#">Выйти</a>
+                            <a href="/logout.php">Выйти</a>
```
1. Из другого задания. 2. logout.php отсутствует

nkolosov commented on templates/layout.php:
```diff
@@ -20,13 +20,13 @@
 
                 <div class="main-header__side">
-                    <a class="main-header__side-item button button--plus open-modal" href="pages/form-task.html">Добавить задачу</a>
+                    <a class="main-header__side-item button button--plus open-modal" href="/add.php">Добавить задачу</a>
```
Тут и в других ссылках - абсолютный путь (с ведущим `/`), это не совсем коректно, путь должен быть относительным, иначе, если проект находится не в корневой директории сервера, работать не будет

nkolosov commented on templates/register.php:
```diff
@@ -0,0 +1,51 @@
+<div class="content">
```
Из другого задания

nkolosov commented on add_project.php:
```diff
@@ -0,0 +1,11 @@
+<?php
```
Из другого задания?

nkolosov commented on templates/main.php:
```diff
@@ -77,15 +77,18 @@
             <!-- Вывод подгружаемого файла в таблицу -->
             <td class="task__file">
-                <a class="download-link" href="#">Home.psd</a>
+                <?php if ($task['path']): ?>
```
Кажется, такого поля нет в БД

nkolosov commented on helpers.php:
```diff
@@ -142,3 +122,45 @@ function include_template($name, array $data = []) {
+    return $count_of_task;
+}
+
+function is_date_correct($date) {
```
Из названия не понятно чем отличаются `is_date_valid` и `is_date_correct`, эту функцию стоит переименовать

nkolosov commented on add.php:
```diff
@@ -0,0 +1,110 @@
+                $rule = $rules[$key];
+                $errors[$key] = $rule($value);
+            }
+            if (in_array($key, $required) && empty($value)) {
```
Логичнее, сначала проверить заполнено ли обязательное поле, а уже потом, если оно заполнено, проверять правильно ли оно заполнено

delannushka committed:
Удалила все лишние страницы, внесла корректировки по функциям.
!! поле path появилось из 32 строки index.php !!
Подкорректировала ссылки на страницы.
